src/db_adapters/assets/non_fungible_token_events.rs

```diff
@@ -0,0 +1,196 @@
+                        interval *= 2;
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn collect_nft_events(
+    outcome: &near_indexer::IndexerExecutionOutcomeWithReceipt,
+    block_timestamp: &u64,
+    shard_id: &near_indexer::near_primitives::types::ShardId,
+    index_in_shard: &mut i32,
+) -> Vec<models::assets::non_fungible_token_events::NonFungibleTokenEvent> {
+    let mut event_logs: Vec<nft_types::Nep171Event> = Vec::new();
+    let prefix = "EVENT_JSON:";
+    for untrimmed_log in &outcome.execution_outcome.outcome.logs {
+        let log = untrimmed_log.trim();
+        if !log.starts_with(prefix) {
+            continue;
+        }
+
+        let event: nft_types::NearEvent = match serde_json::from_str::<'_, nft_types::NearEvent>(
+            log[prefix.len()..].trim(),
+        ) {
+            Ok(result) => result,
+            Err(err) => {
+                warn!(
+                    target: crate::INDEXER_FOR_EXPLORER,
+                    "NFT: provided event log does not correspond to any of formats defined in NEP. Will ignore this event. \n {:#?} \n{:#?}",
+                    err,
+                    untrimmed_log,
+                );
+                continue;
+            }
+        };
+
+        // When other event types will be added, we need to ignore them in this function.
+        // We are interested only in Nep171 events here
+        let nft_types::NearEvent::Nep171(nep171_event) = event;
+        event_logs.push(nep171_event);
+    }
+
+    let mut nft_events = Vec::new();
+    let contract_id = &outcome.receipt.predecessor_id;
+    for log in event_logs {
+        match log.event_kind {
+            nft_types::Nep171EventKind::NftMint(mint_events) => {
+                for mint_event in mint_events {
+                    let memo = mint_event.memo.unwrap_or_else(|| "".to_string());
+                    for token_id in mint_event.token_ids {
+                        nft_events.push(
+                            models::assets::non_fungible_token_events::NonFungibleTokenEvent {
+                                emitted_for_receipt_id: outcome.receipt.receipt_id.to_string(),
+                                emitted_at_block_timestamp: BigDecimal::from(*block_timestamp),
+                                emitted_in_shard_id: BigDecimal::from(*shard_id),
+                                emitted_index_of_event_entry_in_shard: *index_in_shard,
+                                emitted_by_contract_account_id: contract_id.to_string(),
+                                token_id: token_id.to_string(),
+                                event_kind: models::enums::NftEventKind::Mint,
+                                token_old_owner_account_id: "".to_string(),
+                                token_new_owner_account_id: mint_event.owner_id.to_string(),
+                                token_authorized_account_id: "".to_string(),
+                                token_memo: memo.to_string(),
```

Comment:
I don't see a better way either.

P.S. It seems you don't need most of those `.to_string()`, do you?

Reply:
It's possible to emit it only in `token_id`.
In other lines, we have to copy the value, because we use them in cycle and move on each iteration. To say that explicitly, I will change it to `.clone()`